apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/RetryHandler.cs

```diff
@@ -65,5 +79,69 @@ public async Task<bool> HandleResponseAsync(HandleUnsuccessfulResponseArgs args)
             await Task.Delay(delay, args.CancellationToken).ConfigureAwait(false);
             return true;
         }
+
+        /// <summary>
+        /// Interceptor which adds a random invocation ID within the x-goog-api-client header,
+        /// along with an attempt count.
+        /// </summary>
+        private sealed class InvocationIdInterceptor : IHttpExecuteInterceptor
+        {
+            internal static InvocationIdInterceptor Instance { get; } = new InvocationIdInterceptor();
+
+            private const string InvocationIdPrefix = InvocationIdHeaderPart + "/";
+            private const string AttemptCountPrefix = AttemptCountHeaderPart + "/";
+
+            private InvocationIdInterceptor()
+            {
+            }
+
+            public Task InterceptAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+            {
+                // If we don't have the header already, or if there isn't a single value,
+                // that's an odd situation: don't add one with just the invocation ID.
+                if (!request.Headers.TryGetValues(VersionHeaderBuilder.HeaderName, out var values) || values.Count() != 1)
+                {
+                    return Task.CompletedTask;
+                }
+                string value = values.Single();
+                List<string> parts = value.Split(' ').ToList();
+
+                bool gotInvocationId = false;
+                bool gotAttemptCount = false;
+                for (int i = 0; i < parts.Count; i++)
```

Comment:
nit:
```suggestion
                for (int i = 0; i < parts.Count && (!gotInvocationid || !gotAttemptCount); i++)
```

Reply:
Okay, will leave. We're unlikely to add more parts in general - and if we add more parts for retry, this logic would become even hairier (or possibly we'd fail to update it at all...)